server/cluster/cluster.go

```diff
@@ -1349,6 +1349,53 @@ func (c *RaftCluster) SlowStoreRecovered(storeID uint64) {
 	c.core.SlowStoreRecovered(storeID)
 }
 
+// NeedAwakenAllRegionsInStore checks whether we should do AwakenRegions operation.
+func (c *RaftCluster) NeedAwakenAllRegionsInStore(storeID uint64) (needAwaken bool, slowStoreIDs []uint64) {
+	// TODO: Current checking strategy follows the rule that there
+	// is no more than one slowStore in this cluster. We should upate
+	// it in the future if RaftCluster could tolerate several slowStores.
+	var slowStore *core.StoreInfo
+
+	for _, store := range c.GetStores() {
+		if store.IsRemoved() {
+			continue
+		}
+
+		if (store.IsPreparing() || store.IsServing()) && store.IsSlow() {
+			// Do nothing if there is more than one slow store.
+			slowStore = store
+			slowStoreIDs = append(slowStoreIDs, store.GetID())
+		}
+	}
+	store := c.GetStore(storeID)
+	// We will filter out heartbeat requests from slowStores, and just return
+	// AwakenRegions message to those stores in normal state.
+	return slowStore != nil && slowStore.GetStoreStats().GetStoreId() != storeID && store.NeedAwakenStore(), slowStoreIDs
+}
+
+// UpdateAwakenStoreTime updates the last awaken time for the store.
+func (c *RaftCluster) UpdateAwakenStoreTime(storeID uint64, lastAwakenTime time.Time) error {
+	c.Lock()
+	defer c.Unlock()
+
+	store := c.GetStore(storeID)
+	if store == nil {
```

Comment:
if `store == nil || .. ||`?

Reply:
That could save some lines

Reply:
Emmm. IMO, it's better to split them into different lines, because the followings belong to different ErrorTypes.

```Go
        if store == nil {
		return errs.ErrStoreNotFound.FastGenByArgs(storeID) /* A NotFound Error. */
	}

	if store.IsRemoved() {
		return errs.ErrStoreRemoved.FastGenByArgs(storeID) /* A StoreRemoved Error. */
	}

	if store.IsPhysicallyDestroyed() {
		return errs.ErrStoreDestroyed.FastGenByArgs(storeID) /* A StoreDestroyed Error. */
	}
```

Reply:
My fault, I miss read them.